### PR TITLE
All: aws-sdk-v3-js with better Error Handling

### PR DESCRIPTION
**Pull request by leematos**
Thanks to @plerohellec and @spoit for the test buckets on UpCloud and AWS.

These changes here were tested across Linode/AWS/UpCloud and we get (some) clearer errors, but also working sync clients across all of them.

There are some challenges with the aws-js-v3 error handling that we need to decide. Right now, if the HeadBucket request on config check fails with a 301/403, it will not return an error. I believe that issue is: https://github.com/aws/aws-sdk-js-v3/issues/1596. We could fall back to a presigned URL to show the error (lot of work) or right now, if you save that config, and reload the app, the synchronizer will report the error appropriately. In my testing it was always related to a config problem (wrong key/ wrong bucket, etc). 

Things outstanding:

- [x] Update Readme to explain what settings to use (who needs forcePathStyle)
- [x] Run each provider through the test suite
- [ ] Fix any issues that arise in testing.

### CODE REVIEW

**Comment by laurent22** [comment]
There are several conflicts on package-lock files and other files you might want to look at;

**Comment by leematos** [comment]
I've had a chance to test this with all 3 providers (Linode/UpCloud/AWS) and my tests are coming back green.

I'm trying to figure out how to make a test for the fetch command so we have specific s3 error tests. I'm hoping to have that done this weekend.

I need to double check iOS and once I confirm that I'll ask for another android test.
----
The test suite is taking a suuuuuper long time now running in band, and I have to assume it's something to do with this change. @laurent22 are you noticing slowness in your tests? It takes me 22 min on a 2016 MBP to run through in band.

**Comment by leematos** [comment]
I'm gonna get these linter errors in the AM :D

**Comment by laurent22** [comment]
> The test suite is taking a suuuuuper long time now running in band, and I have to assume it's something to do with this change. @laurent22 are you noticing slowness in your tests? It takes me 22 min on a 2016 MBP to run through in band.

That's probably normal. Not only the tests run in band but they are done against a real sync target over the network. This is compared to the normal tests which are run in parallel and using an in-memory sync target.

**Comment by laurent22** [comment]
By the way make sure you have the precommit hook installed as there are many linter errors in the code (see CI).

**Comment by plerohellec** [comment]
@leematos, out of curiosity, in the end was there any difference in the S3 API responses from Linode, UpCloud and AWS?

**Comment by laurent22** [comment]
@leematos, what's the status of this pull request?

**Comment by leematos** [comment]
Apologies, I could have been clearer as to the state. 

I'm mostly reflecting on https://github.com/laurent22/joplin/pull/5312/files#diff-d06c6d5b2a13f7fa8da76bfd1d45a14e77df2d2ad3f92995e7dfe8c13a0616efR261-R287 and if this should/could become a case statement to be clearer.

The error handling as is worked in my testing across AWS/Linode/UpCloud, we just need an android test and then this could get merged or refactored and merged.

Additionally, I've been using the iOS app with v3 api in it without issue, so I'm confident in the API and that it won't cause problems.

@roman-r-m I'm going to send you a bucket, if you could test s3 once you get that, I think we can merge this if there are no errors!

**Comment by leematos** [comment]
Just wanted to follow up that I still intend to finalize this, I am currently dealing with physical-world flooding aftermath in the wake of Hurricane Ida so if I'm silent for the next week or two, that's why 🌦️

**Comment by leematos** [comment]
KK @roman-r-m I'll send you a Linode bucket -- if you could test on android that'd be great!

I've reviewed the code and I think the cases we have here will work, and if not, we'll refactor as we find new error cases.

I'm still using the previously reverted build (with the new AWS SDK) and I have not had any sync problems over the last month of use with near daily use.

**Comment by leematos** [comment]
Roman is hitting some errors now with "SignatureNotMatching" as we test with Android that we are working through.  We aren't yet sure if it's a problem with my S3 creds or code.

**Comment by leematos** [comment]
@roman-r-m I've just updated the branch to include URL on mobile, I don't think that was your problem, but letting you know.

(this conflict blocking the build is _not_ a problem as that's joplin server related)

**Comment by laurent22** [comment]
Ok great, then @leematos if you're happy with it too then we can merge once the conflicts are fixed.

**Comment by laurent22** [comment]
@leematos, let me know if you'd still like to proceed with this pull request. We're probably going to merge it anyway because it's very good, but just want to confirm in case there's something you want to add.

**Comment by leematos** [comment]
Thanks for the follow up and really sorry for the delay! Been meaning to follow up, still dealing with physical world house thangs. 😞 

Everything we have here is tested and should work. I'm comfortable merging.

(If you can update the packages to stop the conflicts that'd be dope, otherwise I can get to it this weekend.)

**Comment by leematos** [comment]
woooosh. I hate lock file merge conflicts 😓 

I _think_ I got this sorted but I need to check after this CI runs.

**Comment by leematos** [comment]
I've reviewed and the packages look right, so the lockfiles should also be good. 

I don't think you have a test flight system @laurent22 but do let me know the day you are targeting releasing this so I can stay alert for any unexpected challenges.

File: package-lock.json
Comment (laurent22) [inline package-lock.json:6334]: Remove file

File: packages/app-clipper/popup/package-lock.json
Comment (laurent22) [inline packages/app-clipper/popup/package-lock.json:7194]: Remove file

File: packages/htmlpack/package-lock.json
Comment (laurent22) [inline packages/htmlpack/package-lock.json:1]: Remove file

File: packages/server/package-lock.json
Comment (laurent22) [inline packages/server/package-lock.json:25736]: Remove file

File: packages/tools/package-lock.json
Comment (laurent22) [inline packages/tools/package-lock.json:19092]: Remove file

**Review by laurent22 — APPROVED** [review]
There are still a few issues with the package-lock files as you've added several that unrelated to your changes. I've marked them all so please remove them.

Also there's still a conflict in app-desktop/package-lock.json probably due to a recent change. You can run `npm install --package-lock-only` from packages/app-desktop to fix it.

File: packages/server/package-lock.json
Comment (leematos) [reply packages/server/package-lock.json:25736]: If I git rm this file I get a conflict.

I literally went to dev branch and copied in what's in dev as of 3 min ago.

I _think_ this is git getting confused because my branch is older so it sees it as a change?

I'd try rebasing, but every time I rebase I get gnarly package-lock conflicts... I feel like I'm the joplin ourobouros

**Comment by leematos** [comment]
> You can run npm install --package-lock-only from packages/app-desktop to fix it.

I've run this, and it hasn't removed the conflict. Rebasing this turns into the nightmare we had above. I'm not sure the simplest way out of it :(

I think there are some package update commits that are making my rebases harder than they should be but I'm not sure.

If you have other guidance happy to try again, but otherwise, if not, please correct the desktop package-lock and this is good to go.

**Comment by laurent22** [comment]
I think rebase brings more problems than it solves in general. Instead you should just merge dev and fix the conflicts. 

I don't know if you've done something or not, but you still have a bunch of unrelated package-lock files in your commit even though they are not relevant, so they should at least be removed. I know dealing with package-lock file is frustrating, it is for me too, but it needs to be done properly. There's a chance you're undoing everything you do when you rebase, so just merge.

It seems if I fix the files myself, I can't push back the commit, so it means I'd need to do it in a separate branch and as a result you won't be the author of the commit anymore. I think it's better if you remain the author.

**Comment by roman-r-m** [comment]
I wonder if you can first resolve the conflict by picking either version, then merge master, and finally re-generate `package-lock.json` files? I'm not an expert on node/npm though so take it with a grain of salt.

**Comment by leematos** [comment]
@roman-r-m I used to rebase on to new dev, get a ton of package file conflicts, regenerate package files finish rebasing and I can get a clean MR, but that's causing too many package files to be updated. I'm gonna talk to some git folks tomorrow at work and see if we can get this clean.

**Comment by leematos** [comment]
With the help of @WarheadsSE I was able to clean up some mess and get my changes applied atop dev.

I ran npm install --package-lock-only in /packages/app-mobile and it updated the appropriate lock files. (not pushed yet) 

I tried the same npm install --package-lock-only inside of lib and it's upset about htmlpack (which is up a dir)

```
lib % npm install --package-lock-only
npm ERR! code E404
npm ERR! 404 Not Found - GET https://registry.npmjs.org/@joplin%2fhtmlpack - Not found
npm ERR! 404 
npm ERR! 404  '@joplin/htmlpack@~2.6' is not in this registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404 
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/lbot/.npm/_logs/2021-11-23T18_05_00_906Z-debug.log
```

If I run npm install at the top level directory I expect to have the same problems as before where it will edit many package locks. I can git add _only_ the mobile and lib package-locks, but I'm trying to avoid making this more of a mess where my local branch gets more merge conflicts.

@laurent22 -- is there another command I should be using or is npm install at the top dir, and only git add the lib / mobile locks the way you'd approach this?

**Comment by laurent22** [comment]
As you've now removed all the package-lock files, you could do like Roman suggested: just merge dev on your branch, then from the root run `npm run clean && npm install`, which should regenerate all the package-lock files.

Npm is probably going to update unrelated lock files, so you need to do `git checkout path/package-lock.json` on these to exclude them from your commit. You should only update the package-lock files that correspond to the package.json you modified.

**Comment by leematos** [comment]
It lives!!!!! 🎉 🎉 🎉 @laurent22 for your hopefully final review. 

If you merge, do let me know when you are targeting an iOS build so I can be on standby.

**Comment by laurent22** [comment]
Looks good now, so let's merge before changes in dev bring again more conflicts! Thanks a lot for your efforts on this @leematos, and thanks @roman-r-m for testing.

**Comment by zxkmm** [comment]
> It lives!!!!! 🎉 🎉 🎉 @laurent22 for your hopefully final review.
> 
> If you merge, do let me know when you are targeting an iOS build so I can be on standby.

Thank you very much!

**Pull request merged**
